From c8ac126d02228b02597c3fc24ea25b97245ff776 Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Tue, 9 Mar 2021 12:36:42 -0800
Subject: [PATCH 1770/1851] drm: xlnx: zynqmp: Ignore DP vblank interrupts if
 external CRTC used

commit b840d5626bd141b8ce1da2f95e2f4535452cd11f from
https://github.com/Xilinx/linux-xlnx.git

Currently vblank interrupts are always handled by the PS DP CRTC. This
causes kernel warnings when an external CRTC is used because the PS DP
CRTC is not registered. This patch checks for an external CRTC and if
one is being used, it will ignore vblank interrupts. Else let the PS DP
CRTC handle the vblank.

Similarly only disable vblank handling when PS DP CRTC is used.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Signed-off-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 3 ++-
 drivers/gpu/drm/xlnx/zynqmp_dp.c   | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index 80cf40cca5c0..a8a4ba1e9740 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -2844,7 +2844,8 @@ zynqmp_disp_crtc_atomic_disable(struct drm_crtc *crtc,
 	zynqmp_disp_clk_disable(disp->pclk, &disp->pclk_en);
 	zynqmp_disp_plane_disable(crtc->primary);
 	zynqmp_disp_disable(disp, true);
-	drm_crtc_vblank_off(crtc);
+	if (!disp->dpsub->external_crtc_attached)
+		drm_crtc_vblank_off(crtc);
 	pm_runtime_put_sync(disp->dev);
 }
 
diff --git a/drivers/gpu/drm/xlnx/zynqmp_dp.c b/drivers/gpu/drm/xlnx/zynqmp_dp.c
index 238cfbfad1e1..0a8b5b9821ad 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_dp.c
@@ -1782,7 +1782,8 @@ static irqreturn_t zynqmp_dp_irq_handler(int irq, void *data)
 	zynqmp_dp_write(dp->iomem, ZYNQMP_DP_SUB_TX_INTR_STATUS, status);
 
 	/* The DP vblank will not be enabled with remote crtc device */
-	if (status & ZYNQMP_DP_TX_INTR_VBLANK_START)
+	if (status & ZYNQMP_DP_TX_INTR_VBLANK_START &&
+		!dp->dpsub->external_crtc_attached)
 		zynqmp_disp_handle_vblank(dp->dpsub->disp);
 
 	if (status & ZYNQMP_DP_TX_INTR_HPD_EVENT)
-- 
2.31.1

